### PR TITLE
flip sprites in combined sequences

### PR DESCRIPTION
**Pull request by evgeniysergeev**
allow to use frames: in Combined: sequences plus allow flip or mirror sprites in sequences.
needed for dune2 mod

### CODE REVIEW

**Comment by evgeniysergeev** [comment]
example yaml:

```
quad.dune2:
	idle: UNITS.SHP
		Combine:
			a: UNITS.SHP
				Frames: 0, 1, 2, 3, 4
				Start: 0
				Length: 5
			b: UNITS.SHP
				Frames: 3, 2, 1
				Start: 1
				Length: 3
				FlipX: true
		Facings: -8
	icon: SHAPES.SHP
		Start: 74
		Offset: 0,0
```

**Comment by evgeniysergeev** [comment]
Flipping will work only then #9723 will be merged.

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 1]
```diff
@@ -131,10 +163,14 @@ public DefaultSpriteSequence(ModData modData, TileSet tileSet, SpriteCache cache
 						// Allow per-sprite offset, start, and length
 						var subStart = LoadField<int>(sd, "Start", 0);
 						var subOffset = LoadField<float2>(sd, "Offset", float2.Zero);
+						bool subFlipX = LoadField<bool>(sd, "FlipX", flipX);
+						bool subFlipY = LoadField<bool>(sd, "FlipY", flipY);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: I think both of these should be `var`.

**Comment by abcdefg30** [comment]
It seems like you didn't use `var` in a lot of places where our code style suggests to use it
(see examle above). Could you change that, please?

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 2]
```diff
@@ -94,6 +95,34 @@ protected static T LoadField<T>(Dictionary<string, MiniYaml> d, string key, T fa
 			return fallback;
 		}
 
+		protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
+		{
+			int l, t, r, b;
+			if (flipX)
```
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: You can make this function much more concise by using terniary statements:
```csharp
var left = flipX ? rect.Right : rect.Left;
var top = flipY ? rect.Bottom : rect.Top;
var right = flipX ? rect.Left : rect.Right;
var bottom = flipY ? rect.Top : rect.Bottom;
return Rectangle.FromLTRB(left, top, right, bottom);
```

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 3]
```diff
@@ -144,6 +180,18 @@ public DefaultSpriteSequence(ModData modData, TileSet tileSet, SpriteCache cache
 							subLength = LoadField<int>(sd, "Length", 1);
 
 						combined = combined.Concat(subSprites.Skip(subStart).Take(subLength));
+
+						int[] frames = LoadField<int[]>(sd, "Frames", null);
```
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: This applies to sub-sequences, so you should be named `subframes`.

Comment (evgeniysergeev) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: what about subStart: and subLength: ?

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 4]
```diff
@@ -144,6 +162,18 @@ public DefaultSpriteSequence(ModData modData, TileSet tileSet, SpriteCache cache
 							subLength = LoadField<int>(sd, "Length", 1);
 
 						combined = combined.Concat(subSprites.Skip(subStart).Take(subLength));
+
+						var subFrames = LoadField<int[]>(sd, "subFrames", null);
```
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: Keep the yaml field as Frames - its only the variable names that we change.

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 5]
```diff
@@ -144,6 +162,18 @@ public DefaultSpriteSequence(ModData modData, TileSet tileSet, SpriteCache cache
...
+						if (subFrames != null)
+						{
+							var fixedFrames = subFrames.Select(i => i + totalCount - subStart).ToArray();
+							if (Frames != null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: This is mixing the overall (after combining) Frames with actions for the sub-frames sequences.  This doesn't look right to me.  The aggregate sequences should be building up a source collection of frames, just like if they were loaded from a single file, and then Frames indexes that later.

**Comment by reaperrr** [comment]
@evgeniysergeev now that #9723 has been merged, could you rebase on bleed? I've done that myself locally, but it would make testing easier for other reviewers.

**Comment by reaperrr** [comment]
The improved Combine can save some yaml duplication, and flipping works fine. :+1: 

My testcase if anyone's interested:
```RA MCV:
mcv:
	idle:
		Combine:
			mcv:
				Length: 17
			mcv:
				Start: 1
				Length: 15
				FlipX: true
		Frames: 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17
		Facings: 32
	icon: mcvicon```

**Comment by evgeniysergeev** [comment]
rebased on bleed

**Comment by penev92** [comment]
:+1:

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/cb55f65042030cef91daf6a071094c29d83702ba)

**Pull request merged**